plugins/net_plugin/net_plugin.cpp

```diff
@@ -2306,17 +2294,40 @@ namespace eosio {
             send_handshakes();
          }
       } else if( state == lib_catchup ) {
-         if( blk_num >= sync_known_lib_num ) {
+         std::unique_lock g_sync( sync_mtx );
+         if( blk_applied && blk_num >= sync_known_lib_num ) {
             peer_dlog( c, "All caught up with last known last irreversible block resending handshake" );
             set_state( in_sync );
             g_sync.unlock();
             send_handshakes();
-         } else if( blk_num >= sync_last_requested_num ) {
-            request_next_chunk( std::move( g_sync) );
          } else {
-            g_sync.unlock();
-            peer_dlog( c, "calling sync_wait" );
-            c->sync_wait();
+            if (!blk_applied) {
+               if (blk_num >= c->sync_last_requested_block) {
+                  peer_dlog(c, "calling cancel_wait, block ${b}", ("b", blk_num));
+                  c->cancel_wait();
+               } else {
+                  peer_dlog(c, "calling sync_wait, block ${b}", ("b", blk_num));
+                  c->sync_wait();
+               }
+
+               sync_next_expected_num = blk_num + 1;
+            } else {
+               if (sync_last_requested_num != 0 && blk_num >= sync_last_requested_num) {
+                  sync_next_expected_num = blk_num + 1;
+                  request_next_chunk(std::move(g_sync));
+                  return;
+               }
+            }
+
+            uint32_t head = my_impl->get_chain_head_num();
+            if (head + sync_req_span > sync_last_requested_num) { // don't allow to get too far head (one sync_req_span)
+               if (sync_next_expected_num > sync_last_requested_num && sync_last_requested_num < sync_known_lib_num) {
+                  fc_dlog(logger, "Requesting range ahead, head: ${h} blk_num: ${bn} sync_next_expected_num ${nen} sync_last_requested_num: ${lrn}",
+                          ("h", head)("bn", blk_num)("nen", sync_next_expected_num)("lrn", sync_last_requested_num));
+                  request_next_chunk(std::move(g_sync));
+               }
+            }
+
```

Comment:
I see two issues:
- `>=` at line 2315 should be `>`
- missing `sync_next_expected_num = blk_num + 1;` after like 2319

I think this code could be updated to:

```suggestion
            sync_next_expected_num = blk_num + 1;
            if (!blk_applied) {
               g_sync.unlock();
               if (blk_num >= c->sync_last_requested_block) {
                  peer_dlog(c, "calling cancel_wait, block ${b}", ("b", blk_num));
                  c->cancel_wait();
               } else {
                  peer_dlog(c, "calling sync_wait, block ${b}", ("b", blk_num));
                  c->sync_wait();
               }
            } else {
               if (sync_last_requested_num != 0) {
                  if (blk_num + sync_req_span > sync_last_requested_num &&
                      sync_next_expected_num > sync_last_requested_num &&
                      sync_last_requested_num < sync_known_lib_num) {
                     fc_dlog(logger, "Requesting range ahead, head: ${h} blk_num: ${bn} sync_next_expected_num ${nen} sync_last_requested_num: ${lrn}",
                             ("h", head)("bn", blk_num)("nen", sync_next_expected_num)("lrn", sync_last_requested_num));
                     request_next_chunk(std::move(g_sync));
                  }
               }
            }```

Reply:
`sync_recv_block` is called when a block is received (`blk_applied` == `false`) and when a block is applied (`blk_applied` == `true`). 

We should probably have better handling of `sync_last_requested_num == 0` in all cases; I'll fix that up better.

As for the `>=` on line 2315. If it was `>` only then it would not be hit. You will not receive the block after `sync_last_requested_num` unless you request it.

Reply:
You don't want to `sync_next_expected_num = blk_num + 1` when a block is applied because that is behind where you are receiving blocks. So you only increment that when receiving blocks.

I do have what might be nothing more than a safety net in if the applied block is the `sync_last_requested_num` to attempt to request another range.